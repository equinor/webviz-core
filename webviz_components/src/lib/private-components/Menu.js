/* eslint-disable react/no-children-prop */
/* eslint-disable react/no-unused-prop-types */
import React, {Component} from 'react';
import PropTypes from 'prop-types';
import {Route, Link} from 'react-router-dom';
import styled, {css} from 'styled-components';
import MenuButton from './MenuButton';

const OuterWrapper = styled.div`
    background: var(--menuBackground);
    @media (min-width: 1201px) {
        min-height: fill-available;
    }
`;

const displayNone = css`
    display: none;
`;

const Wrapper = styled.div`
    background: var(--menuBackground);
    min-width: 250px;
    display: flex;
    flex-direction: column;
`;

const SubLinksWrapper = styled.div`
    display: flex;
    flex-direction: column;
    @media (max-width: 1200px) {
        ${({menuOpen}) => !menuOpen && displayNone};
    }
    @media (min-width: 1201px) {
        height: 100%;
    }
`;

const MainLinkWrapper = styled.div`
    display: flex;
    align-items: center;
`;

const LinkHome = styled(Link)`
    text-decoration: none;
    font-size: 30px;
    font-family: system-ui;
    @media (max-width: 1200px) {
        text-align: center;
        margin-left: -50px;
    }
`;

const LogoWrapper = styled.div`
    margin: 12px 20px 20px 18px;
    @media (max-width: 1200px) {
        height: 50px;
        margin: 15px;
        display: flex;
        justify-content: center;
        align-items: center;
        width: 100%;
    }
    margin: 20px 30px;
    display: flex;
    align-items: center;
`;

const selectedStyle = css`
    background: var(--menuLinkBackgroundSelected);
    color: var(--menuLinkColorSelected);
`;

const hoverStyle = css`
    background: var(--menuLinkBackgroundHover);
    color: var(--menuLinkHoverColor);
`;

const SubPageLink = styled(Link)`
    text-decoration: none;
    padding: 20px 30px;
    font-size: 16px;
    font-family: var(--menuLinkFont);
    font-weight: var(--menuLinkFontWeight);
    font-style: var(--menuLinkFontStyle);
    border-bottom: 1px solid white;
    color: var(--menuLinkColor);
    :hover {
        ${({selected}) => !selected && hoverStyle};
    }
    ${({selected}) => selected && selectedStyle};
`;

const ButtonWrapper = styled.div`
    @media (min-width: 1201px) {
        display: none;
    }
`;

class Menu extends Component {
    constructor(props) {
        super(props);
        this.state = {
            menuOpen: false,
        };
    }

    render() {
        /* eslint-disable react/destructuring-assignment, react/prop-types */
        const {props} =
            this.props._dashprivate_layout &&
            this.props._dashprivate_layout.props
                ? this.props._dashprivate_layout
                : this;
        /* eslint-enable react/destructuring-assignment, react/prop-types */

        const {subPages, basePath} = props;
        const {menuOpen} = this.state;

        return (
            <OuterWrapper>
                <Wrapper>
                    <MainLinkWrapper>
                        <ButtonWrapper>
                            <MenuButton
                                onClick={() => {
                                    this.setState({menuOpen: !menuOpen});
                                }}
                            />
                        </ButtonWrapper>
                        <Route
                            exact
                            path={`${basePath}`}
                            children={() => (
                                <LogoWrapper>
                                    <LinkHome to={`${basePath}`}>
                                        <div id="logo" />
                                    </LinkHome>
                                </LogoWrapper>
                            )}
                        />
                    </MainLinkWrapper>
                    <SubLinksWrapper menuOpen={menuOpen}>
                        {subPages.map(({title, id}) => (
                            <Route
                                key={`subpage-${id}`}
                                path={`${basePath}/${id}`}
                                children={({match}) => (
                                    <SubPageLink
                                        selected={match}
                                        to={`${basePath}/${id}`}
                                    >
                                        {title}
                                    </SubPageLink>
                                )}
                            />
                        ))}
                    </SubLinksWrapper>
                </Wrapper>
            </OuterWrapper>
        );
    }
}

Menu.defaultProps = {
    subPages: [],
    basePath: '',
};

Menu.propTypes = {
    subPages: PropTypes.arrayOf(
        PropTypes.shape({
            id: PropTypes.string.isRequired,
            title: PropTypes.string.isRequired,
        })
    ),
    basePath: PropTypes.string,
};

export default Menu;
