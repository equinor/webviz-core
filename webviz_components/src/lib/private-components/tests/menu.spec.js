import React from 'react';
import renderer from 'react-test-renderer';
import {BrowserRouter} from 'react-router-dom';
import Menu from '../Menu';

describe('Menu', () => {
    it('should render Menu', () => {
        const tree = renderer.create(
            <BrowserRouter>
                <Menu
                    subPages={[
                        {
                            id: 'subpage 1',
                            title: 'Subpage 1',
                        },
                        {
                            id: 'subpage 2',
                            title: 'Subpage 2',
                        },
                    ]}
                />
            </BrowserRouter>
        );
        expect(tree.toJSON()).toMatchSnapshot();
    });
});
