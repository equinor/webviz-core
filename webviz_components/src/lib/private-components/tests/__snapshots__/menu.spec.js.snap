// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Menu should render Menu 1`] = `
.c4 {
  color: var(--menuLinkColor);
  height: 50px;
  width: 50px;
  border: none;
  background: transparent;
  outline: none;
  cursor: pointer;
}

.c4:hover {
  color: var(--menuLinkHoverColor);
}

.c0 {
  background: var(--menuBackground);
}

.c1 {
  background: var(--menuBackground);
  min-width: 250px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c7 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c6 {
  -webkit-text-decoration: none;
  text-decoration: none;
  font-size: 30px;
  font-family: system-ui;
}

.c5 {
  margin: 12px 20px 20px 18px;
  margin: 20px 30px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c8 {
  -webkit-text-decoration: none;
  text-decoration: none;
  padding: 20px 30px;
  font-size: 16px;
  font-family: var(--menuLinkFont);
  font-weight: var(--menuLinkFontWeight);
  font-style: var(--menuLinkFontStyle);
  border-bottom: 1px solid white;
  color: var(--menuLinkColor);
}

.c8:hover {
  background: var(--menuLinkBackgroundHover);
  color: var(--menuLinkHoverColor);
}

@media (min-width:1201px) {
  .c0 {
    min-height: -webkit-fill-available;
    min-height: -moz-available;
    min-height: fill-available;
  }
}

@media (max-width:1200px) {
  .c7 {
    display: none;
  }
}

@media (min-width:1201px) {
  .c7 {
    height: 100%;
  }
}

@media (max-width:1200px) {
  .c6 {
    text-align: center;
    margin-left: -50px;
  }
}

@media (max-width:1200px) {
  .c5 {
    height: 50px;
    margin: 15px;
    display: -webkit-box;
    display: -webkit-flex;
    display: -ms-flexbox;
    display: flex;
    -webkit-box-pack: center;
    -webkit-justify-content: center;
    -ms-flex-pack: center;
    justify-content: center;
    -webkit-align-items: center;
    -webkit-box-align: center;
    -ms-flex-align: center;
    align-items: center;
    width: 100%;
  }
}

@media (min-width:1201px) {
  .c3 {
    display: none;
  }
}

<div
  className="c0"
>
  <div
    className="c1"
  >
    <div
      className="c2"
    >
      <div
        className="c3"
      >
        <button
          className="c4"
          onClick={[Function]}
        >
          <svg
            fill="None"
            height="18px"
            stroke="currentColor"
            strokeLinecap="round"
            strokeLinejoin="round"
            strokeWidth="4"
            width="22px"
          >
            <path
              d="M2,2 L20,2 M2,9 L20,9 M2,16 L20,16"
            />
          </svg>
        </button>
      </div>
      <div
        className="c5"
      >
        <a
          className="c6"
          href="/"
          onClick={[Function]}
        >
          <div
            id="logo"
          />
        </a>
      </div>
    </div>
    <div
      className="c7"
    >
      <a
        className="c8"
        href="/subpage 1"
        onClick={[Function]}
        selected={null}
      >
        Subpage 1
      </a>
      <a
        className="c8"
        href="/subpage 2"
        onClick={[Function]}
        selected={null}
      >
        Subpage 2
      </a>
    </div>
  </div>
</div>
`;
